Modules/_ctypes/stgdict.c

```diff
@@ -411,231 +372,63 @@ PyCStructUnionType_update_stginfo(PyObject *type, PyObject *fields, int isStruct
         }
         ffi_ofs = baseinfo->length;
     } else {
-        offset = 0;
-        size = 0;
-        align = 0;
-        union_size = 0;
-        total_align = forced_alignment;
         stginfo->ffi_type_pointer.type = FFI_TYPE_STRUCT;
         stginfo->ffi_type_pointer.elements = PyMem_New(ffi_type *, len + 1);
         if (stginfo->ffi_type_pointer.elements == NULL) {
             PyErr_NoMemory();
-            return -1;
+            goto error;
         }
         memset(stginfo->ffi_type_pointer.elements, 0,
                sizeof(ffi_type *) * (len + 1));
         ffi_ofs = 0;
     }
 
-    assert(stginfo->format == NULL);
-    if (isStruct) {
-        stginfo->format = _ctypes_alloc_format_string(NULL, "T{");
-    } else {
-        /* PEP3118 doesn't support union. Use 'B' for bytes. */
-        stginfo->format = _ctypes_alloc_format_string(NULL, "B");
-    }
-    if (stginfo->format == NULL)
-        return -1;
-
     for (i = 0; i < len; ++i) {
-        PyObject *name = NULL, *desc = NULL;
-        PyObject *pair = PySequence_GetItem(fields, i);
-        PyObject *prop;
-        Py_ssize_t bitsize = 0;
-
-        if (!pair || !PyArg_ParseTuple(pair, "UO|n", &name, &desc, &bitsize)) {
-            PyErr_SetString(PyExc_TypeError,
-                            "'_fields_' must be a sequence of (name, C type) pairs");
-            Py_XDECREF(pair);
-            return -1;
+        PyObject *prop_obj = PyTuple_GET_ITEM(layout_fields, i);
+        assert(prop_obj);
+        if (!PyType_IsSubtype(Py_TYPE(prop_obj), st->PyCField_Type)) {
+            PyErr_Format(PyExc_TypeError,
+                        "fields must be of type CField, got %T", prop_obj);
+            goto error;
+
         }
-        if (PyCArrayTypeObject_Check(st, desc)) {
+        CFieldObject *prop = (CFieldObject *)prop_obj; // borrow from prop_obj
+
+        if (prop->index != i) {
+            PyErr_Format(PyExc_ValueError,
+                        "field %R index mismatch (expected %d, got %d)",
+                         prop->name, i, prop->index);
```

Comment:
```suggestion
            PyErr_Format(PyExc_ValueError,
                         "field %R index mismatch (expected %ud, got %zd)",
                          prop->name, i, prop->index);
```

By the way, can we avoid having `Py_ssize_t i` at the very top of the function? While I know that it's ANSI C, we are not required to do it anymore so if we can avoid it, it would be easier (I needed to scroll to the top to check the type).

Reply:
And it's not "ud", it should "zd" (typo in my suggestion)